File: snippets/ts/src/use-cases/microservices/order-service.ts
import * as restate from "@restatedev/restate-sdk";

type Order = {
  id: string;
  items: { id: string; quantity: number }[];
  creditCard: string;
};

function chargePayment(creditCard: string, paymentId: string) {
  return undefined;
}

function reserveInventory(id: string, quantity: number) {
  return undefined;
}

// <start_here>
export const orderService = restate.service({
  name: "OrderService",
  handlers: {
    process: async (ctx: restate.Context, order: Order) => {
      // Each step is automatically durable and resumable
      const paymentId = ctx.rand.uuidv4();

      const payment = await ctx.run("charge", () =>
        chargePayment(order.creditCard, paymentId)
      );

      for (const item of order.items) {
        await ctx.run(`reserve-${item.id}`, () =>
          reserveInventory(item.id, item.quantity)
        );
      }

      return { success: true, paymentId };
    },
  },
});
// <end_here>

File: snippets/java/src/main/java/usecases/microservices/OrderService.java
package usecases.microservices;

import static usecases.microservices.utils.Utils.chargePayment;
import static usecases.microservices.utils.Utils.reserveInventory;

import dev.restate.sdk.Restate;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import java.util.UUID;
import usecases.microservices.utils.Order;
import usecases.microservices.utils.OrderResult;

// <start_here>
@Service
public class OrderService {

  @Handler
  public OrderResult process(Order order) {
    // Each step is automatically durable and resumable
    String paymentId = UUID.randomUUID().toString();

    Restate.run("charge-payment", () -> chargePayment(order.creditCard, paymentId));

    for (var item : order.items) {
      Restate.run("reserve-inventory", () -> reserveInventory(item.id, item.quantity));
    }

    return new OrderResult(true, paymentId);
  }
}
// <end_here>
